docs/specs/m2m.md

```diff
@@ -0,0 +1,1218 @@
+CREATE UNIQUE INDEX _auth_client_resource_scope_unique ON _auth_client_resource USING btree (app_id, client_id, resource_id, scope_id);
+
+-- A sibling table of _auth_oauth_authorization, that takes resource_id into account.
+CREATE TABLE _auth_oauth_authorization_resource (
+  id text PRIMARY KEY,
+  app_id text NOT NULL,
+  client_id text NOT NULL,
+  user_id text NOT NULL REFERENCES _auth_user(id),
+  resource_id text NOT NULL REFERENCES _auth_resource(id),
+  scope_id text NOT NULL REFERENCES _auth_resource_scope(id)
+)
+CREATE UNIQUE INDEX _auth_oauth_authorization_resource_unique ON _auth_oauth_authorization_resource USING btree (app_id, client_id, user_id, resource_id, scope_id);
+```
+
+### Changes in Admin API
+
+The changes are mainly the CRUD of Resources and Scopes.
+
+The CRUD of Resources and Scopes DO NOT generate events.
+
+The following GraphQL schema snippet describe the changes to the Admin API GraphQL schema.
+
+```graphql
+type Query {
+  """If clientID is null, then all resources are returned in a paginated fashion."""
+  """If clientID is specified, then all resources associated with the clientID are returned in a paginated fashion."""
+  """If searchKeyword is non-null, a prefix search of uri or name is performed."""
+  """If both clientID and searchKeyword are specified, they are AND-ed."""
+  resources(clientID: String, searchKeyword: String, after: String, before: String, first: Int, last: Int): ResourceConnection
+}
+
+type Mutation {
+  createResource(input: CreateResourceInput!): CreateResourcePayload!
+  updateResource(input: UpdateResourceInput!): UpdateResourcePayload!
+  deleteResource(input: DeleteResourceInput!): DeleteResourcePayload!
+
+  createScope(input: CreateScopeInput!): CreateScopePayload!
+  updateScope(input: UpdateScopeInput!): UpdateScopePayload!
+  deleteScope(input: DeleteScopeInput!): DeleteScopePayload!
+
+  addResourceToClientID(input: AddResourceToClientIDInput!): AddResourceToClientIDPayload!
+  removeResourceFromClientID(input: RemoveResourceFromClientIDInput!): RemoveResourceFromClientIDPayload!
+  addScopesToClientID(input: AddScopesToClientIDInput!): AddScopesToClientIDPayload!
+  removeScopesFromClientID(input: RemoveScopesFromClientIDInput!): RemoveScopesFromClientIDPayload!
+  replaceScopesOfClientID(input: ReplaceScopesOfClientIDInput!): ReplaceScopesOfClientIDPayload!
+}
+
+type Resource implements Entity & Node {
+  id: ID!
+  createdAt: DateTime!
+  updatedAt: DateTime!
+  uri: String!
+  name: String
+  """If clientID is null, then all scopes of this Resource is returned."""
+  """If clientID is specified, then only scopes that are associated with clientID is returned."""
+  """If searchKeyword is non-null, a prefix search of scope is performed."""
+  """If both clientID and searchKeyword are specified, they are AND-ed."""
+  scopes(clientID: String, searchKeyword: String, after: String, before: string, first: Int, last: Int): ScopeConnection
```

Comment:
@louischan-oursky I am thinking should we remove the pagination parameters from this query, and always return all scopes under a resource, because looking at the portal UI design makes me thinks that actually we want to get all scopes in most of the cases.

Reply:
@tung2744 @fungc-io Per our discussion in the product meeting, let's place a hard limit of the number of resources and scopes.

1. A project can only have at most 10000 resources. `apierrors.Forbidden.WithReason("ResourceQuotaExceeded")`
2. A resource can only have at most 10000 scopes. `apierrors.Forbidden.WithReason("ScopeQuotaExceeded")`